elide-core/src/main/java/com/yahoo/elide/security/PermissionManager.java

```diff
@@ -0,0 +1,659 @@
+     * @param changeSpec ChangeSpec
+     * @param <A> type parameter
+     * @see com.yahoo.elide.annotation.CreatePermission
+     * @see com.yahoo.elide.annotation.ReadPermission
+     * @see com.yahoo.elide.annotation.UpdatePermission
+     * @see com.yahoo.elide.annotation.DeletePermission
+     */
+    public <A extends Annotation> void checkPermission(Class<A> annotationClass,
+                                                       PersistentResource resource,
+                                                       ChangeSpec changeSpec) {
+        A annotation = resource.getDictionary().getAnnotation(resource, annotationClass);
+
+        if (annotation == null) {
+            return;
+        }
+
+        PermissionManager.ExtractedChecks extracted = PermissionManager.extractChecks(annotationClass, annotation);
+        CheckMode mode = extracted.getCheckMode();
+        Class<OperationCheck>[] opChecks = extracted.getOperationChecks();
+        Class<CommitCheck>[] comChecks = extracted.getCommitChecks();
+
+        boolean isUpdate = UpdatePermission.class.isAssignableFrom(annotationClass);
+
+        try {
+            runPermissionChecks(opChecks, mode, resource, changeSpec, isUpdate);
+        } catch (ForbiddenAccessException e) {
+            if (mode == CheckMode.ALL || comChecks.length < 1) {
+                log.debug("Forbidden access at entity-level.");
+                throw e;
+            }
+        }
+
+        // If that succeeds, queue up our commit checks
+        if (!isEmptyCheckArray(comChecks)) {
+            commitChecks.add(() -> {
+                runPermissionChecks(comChecks, mode, resource, changeSpec, isUpdate);
+                return null;
+            });
+        }
+    }
+
+    /**
+     * Check for permissions on a class and its fields.
+     *
+     * @param resource resource
+     * @param changeSpec change spec
+     * @param annotationClass annotation class
+     * @param <A> type parameter
+     */
+    public <A extends Annotation> void checkFieldAwarePermissions(PersistentResource<?> resource,
+                                                                  ChangeSpec changeSpec,
+                                                                  Class<A> annotationClass) {
+        checkFieldAwarePermissions(resource, changeSpec, annotationClass, null);
+    }
+
+    /**
+     * Check for permissions on a specific field.
+     *
+     * @param resource resource
+     * @param changeSpec changepsec
+     * @param annotationClass annotation class
+     * @param field field to check
+     * @param <A> type parameter
+     */
+    public <A extends Annotation> void checkFieldAwarePermissions(PersistentResource<?> resource,
+                                                                  ChangeSpec changeSpec,
+                                                                  Class<A> annotationClass,
+                                                                  String field) {
```

Comment:
I don't think we should pass in null here for field to mean anything.  I think we should just have a separate function that doesn't take a field (which implies all fields).